src/relay/pass/quantize.cc

```diff
@@ -0,0 +1,420 @@
+  const auto& cfg = QConfig::Current();
+  Expr data = this->data;
+  if (cfg->store_lowbit_output) {
+    data = Cast(data, Int(8));
+  }
+  // dequantize
+  data = Cast(data, Float(32));
+  data = Multiply(data, this->dom_scale);
+  return data;
+}
+
+QRealizeIntExpr QRealizeIntExprNode::make(Expr data, Expr dom_scale, int nbit, DataType dtype) {
+  NodePtr<QRealizeIntExprNode> n = make_node<QRealizeIntExprNode>();
+  n->data = std::move(data);
+  n->dom_scale = std::move(dom_scale);
+  n->nbit = std::move(nbit);
+  n->dtype = std::move(dtype);
+  return QRealizeIntExpr(n);
+}
+
+
+inline Expr ForwardOp(const Call& ref_call, const Array<Expr>& args) {
+  return CallNode::make(ref_call->op,
+    args, ref_call->attrs, ref_call->type_args);
+}
+
+
+/* calculate `data * s1 / s2`, use shift if possible */
+inline Expr MulAndDiv(Expr data, float s1, float s2) {
+  float shift_factor = std::log2(s1 / s2);
+  if (static_cast<int>(shift_factor) == shift_factor) {
+    return LeftShift(data, MakeConstantScalar(Int(32), static_cast<int>(shift_factor)));
+  } else {
+    data = Cast(data, Float(32));
+    return Multiply(data, MakeConstantScalar(Float(32), s1 / s2));
+  }
+}
+
+Expr QuantizeRealize(const Call& ref_call,
+                     const Array<Expr>& new_args,
+                     const NodeRef& ctx) {
+  const QConfig& cfg = QConfig::Current();
+  // do not handle data type cast
+  const auto param = ref_call->attrs.as<SimulatedQuantizeAttrs>();
+  CHECK_EQ(param->rounding, "round");
+
+  Expr dom_scale = new_args[1];
+  Expr bit = new_args[2];
+  Expr clip_min = new_args[3];
+  Expr clip_max = new_args[4];
+
+  float dom_scale_imm = GetScalarFromConstant<float>(dom_scale);
+  int bit_imm = GetScalarFromConstant<int>(bit);
+  float clip_min_imm = GetScalarFromConstant<float>(clip_min);
+  float clip_max_imm = GetScalarFromConstant<float>(clip_max);
+
+  // x * idom_scale = y * odom_scale
+  // => y = x * idom_scale / odom_scale
+  if (const auto* n = new_args[0].as<QRealizeIntExprNode>()) {
+    Expr data = n->data;
+    float idom_scale_imm = GetScalarFromConstant<float>(n->dom_scale);
+    float odom_scale_imm = GetScalarFromConstant<float>(dom_scale);
+    float shift_nbit = std::log2(odom_scale_imm / idom_scale_imm);
+    // int32->int8
+    if (static_cast<int>(shift_nbit) == shift_nbit) {
+      // use shift
+      if (cfg->round_for_shift) {
+        float round_bias = std::pow(2, shift_nbit - 1);
+        data = Add(data, MakeConstantScalar(Int(32), static_cast<int>(round_bias)));
+      }
+      data = RightShift(data, MakeConstantScalar(Int(32), static_cast<int>(shift_nbit)));
+      float clip_threshold = std::pow(2, bit_imm - 1) - 1;
+      data = Clip(data, - clip_threshold, clip_threshold);
+      return QRealizeIntExprNode::make(data, dom_scale, bit_imm, n->dtype);
+    } else {
+      // float computation
+      data = Cast(data, Float(32));
+      Expr scaled_data = Multiply(data, Divide(n->dom_scale, dom_scale));
```

Comment:
This caused segmentatiuon fault. The rhs of multiply is a const scalar (the ir is like `%8 = multiply(%7, 16f) # ty=Tensor[(1, 64, 224, 224), float32]`).
In the runtime, the host side allocates a device tensor of shape (1,) and initializes with the scalar value, which leads to segmentation fault.

Possible actions:
1. Pass scalar value instead of placeholder to topi compute. This is currently impossible because `FTVMCompute` only accepts tensor argument.
2. Pass scalar argument by value to kernel.

cc @tqchen

Reply:
@vinx13 this sounds like a bug in the compiler, currently, constant scalar value should be inlined directly into the code when possible. Even if we allocate a device scalar, the initialization code should use NDArray::CopyFrom instead of directly de-reference, can you look a bit into it and point out the place that need to be fixed?

Reply:
@tqchen Using injective tag does not work because `traverse_inline` only inline compute between output and the scheduled op (like conv2d, multiply). We need to explicitly call `compute_inline`. https://github.com/dmlc/tvm/commit/37bd9ba4a04885014eef96aa8a818588a11ba073 fixed this issue.

Reply:
@vinx13 nice catch, please send a PR upstream.

Reply:
@vinx13 Nice catch! Thank you!